### ISSUE TITLE
Salt provisioner fails to upload minion config

### ISSUE THREAD

**Issue by piohhmy** [mitchellh/vagrant#5973]
Ran into a problem in 1.7.3 doing a vagrant up using a trusty64 box with a salt provisioner specifying a custom minion_config. This is the error:
```
Copying salt minion config to /etc/salt
Failed to upload a file to the guest VM via SCP due to a permissions
error. This is normally because the SSH user doesn't have permission
to write to the destination location. Alternately, the user running
Vagrant on the host machine may not have permission to read the file.

Source: /Users/danny/dev/se/one-offs/vagrant/minion.yaml
Dest: /etc/salt/minion
```

It appears that the vagrant user does not have permission by default to scp to /etc/salt/ so provisioning fails.  I confirmed it is possible to vagrant ssh into the machine and change the permissions/owner on /etc/salt/ and then the provisioning will work OK.  

It appears 74d22069d41b35bae50e32c1c68a92a4958afebb may have been the breaking change as it introduced the uploading to /etc/salt.

**Comment by SoftCreatR** [comment]
I can confirm this. We're currently experiencing this issue, too.

**Comment by sephlaire** [comment]
I can confirm this as well. I'm using Vagrant 1.7.3 with VirtualBox 5.0. As a workaround I've been using a shell provisioner with the following to make sure salt in installed and the config is copied over:

```
#!/usr/bin/env bash

echo "Setup saltstack PPA"
sudo apt-get install software-properties-common python-software-properties -y >> /dev/null
sudo add-apt-repository ppa:saltstack/salt -y >> /dev/null
sudo apt-get update >> /dev/null

echo "Installing salt-minion"
sudo apt-get install salt-minion -y >> /dev/null
sudo cp /vagrant/salt/configs/ubuntu.conf /etc/salt/minion

echo "Restart salt-minion"
sudo salt-minion -d
sudo service salt-minion restart
```

**Comment by nmadhok** [comment]
I can confirm this issue. As a workaround, I specified ``config.ssh.username = "root"`` and ``config.ssh.password = "vagrant"`` in my Vagrant file

**Comment by clamor** [comment]
Same issue with provider vmware_desktop.

**Comment by royka** [comment]
I've got this issue too

**Comment by amchoukir** [comment]
I can confirm the issue too

**Comment by sonoranDesert** [comment]
This also started happening to me after upgrading to Vagrant 1.7.4.

**Comment by dwickwire** [comment]
Same experience here using vagrant 1.7.3 and 1.7.4 on both Virtual Box 4.3 and 5 using OSX Yosemite.

**Comment by jake223** [comment]
@sephlaire could you explain your workaround in more detail?

**Comment by sephlaire** [comment]
@jake223 
I have 3 relevant files that I am using to provision the VM. `Vagrantfile`, `server.yaml`, and `ubuntu_bootstrap.sh` which are below. The Vagrantfile is using `vm.provision "shell"` to call the shell script which installs salt and copies over my minion config.  
  
##### Vagrantfile
```ruby
# -*- mode: ruby -*-
# # vi: set ft=ruby :

# Specify minimum Vagrant version and Vagrant API version
Vagrant.require_version ">= 1.6.0"
VAGRANTFILE_API_VERSION = "2"

# Require YAML module
require 'yaml'

# Read YAML file with box details
servers = YAML.load_file('servers.yaml')

# Create boxes
Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|

  # Iterate through entries in YAML file
  servers.each do |servers|
    config.vm.define servers["name"] do |srv|
      srv.vm.box = servers["box"]
      srv.vm.network "private_network", ip: servers["ip"]
      srv.vm.hostname = servers["hostname"]
      srv.vm.provision "shell", path: servers["bootstrap"] unless servers["bootstrap"].nil?
      srv.vm.network "forwarded_port", guest: servers["guest_port"], host: servers["host_port"] unless servers["guest_port"].nil? || servers["host_port"].nil?
      srv.vm.synced_folder servers["sync-local"], servers["sync-remote"] unless servers["sync-local"].nil? || servers["sync-remote"].nil?
      srv.vm.provider :virtualbox do |vb|
        vb.name = servers["name"]
        vb.memory = servers["ram"]
      end
      unless servers["minion_config"].nil?
       # srv.vm.provision "file", source: servers["minion_config"], destination: "/etc/salt/minion"
        srv.vm.provision :salt do |salt|
          salt.run_highstate = true
          salt.minion_config = servers["minion_config"] 
          salt.minion_key = servers["minion_key"] unless servers["minion_key"].nil?
          salt.minion_pub = servers["minion_pub"] unless servers["minion_pub"].nil?
        end
      end
    end
  end
end
```

##### servers.yaml
```YAML
---
- name: ubuntu
  hostname: ubuntu
  box: ubuntu/trusty64
  ram: 1024
  ip: 10.1.15.11
  bootstrap: ubuntu_bootstrap.sh
  sync-local: ./salt/roots
  sync-remote: /srv/salt
  minion_config: ./salt/configs/ubuntu.conf
  minion_key: ./salt/ssh-key/ubuntu.pem
  minion_pub: ./salt/ssh-key/ubuntu.pub
- name: centos
  hostname: centos
  box: chef/centos-7.0
  ram: 1024
  ip: 10.1.15.12
  bootstrap: centos_bootstrap.sh
  sync-local: ./salt/roots
  sync-remote: /srv/salt 
  minion_config: ./salt/configs/centos.conf
  minion_key: ./salt/ssh-key/centos.pem
  minion_pub: ./salt/ssh-key/centos.pub
```

##### ubunt_bootstrap.sh
```shell
#!/usr/bin/env bash

echo "Setup saltstack PPA"
sudo apt-get install software-properties-common python-software-properties -y >> /dev/null
sudo add-apt-repository ppa:saltstack/salt -y >> /dev/null
sudo apt-get update >> /dev/null

echo "Installing salt-minion"
sudo apt-get install salt-minion -y >> /dev/null
sudo cp /vagrant/salt/configs/ubuntu.conf /etc/salt/minion

echo "Restart salt-minion"
sudo salt-minion -d
sudo service salt-minion restart
sudo salt-call state.highstate
```

**Comment by amchoukir** [comment]
The workaround that worked for me is as follows

In `Vagrantfile`:

```
config.vm.provision :shell, path: "provision/base_salt_provision.sh", privileged: false
```

In `provision/base_salt_provision.sh`:

```
start_seconds="$(date +%s)"

if [ ! -d "/etc/salt" ]; then
    echo "Bootstrap salt"
    wget -O - https://bootstrap.saltstack.com | sudo sh
    sudo cp /vagrant/salt/minion /etc/salt/minion

    echo "Restart salt-minion"
    sudo salt-minion -d
    sudo service salt-minion restart
fi

echo "Executing salt highstate (provisionning)"
sudo salt-call state.highstate

end_seconds="$(date +%s)"
echo "-----------------------------"
echo "Provisioning complete in "$(expr $end_seconds - $start_seconds)" seconds"
```

**Comment by CharlesLovely** [comment]
Also seeing the same issue.

Also, to add to this current issue, I noticed that when provisioning salt-master, while vagrant won't necessarily error when specifying a configuration file, it won't use the specified file either.

For those looking for yet another workaround, this is what I did:
```bash

# SALT MASTER

saltmaster.vm.provision :salt do |salt|
  salt.master_config = "salt/configs/master"  # this line seems ignored, or may be failing silently
  salt.install_master = true
  salt.no_minion = true
  salt.install_type = "git"
  salt.install_args = "v2015.5.1"
end  # salt provisioner

saltmaster.vm.provision "shell",
  inline: "sudo cp /vagrant/salt/configs/master /etc/salt/master && sudo service salt-master restart"

# SALT MINION

machine.vm.provision :salt do |salt|
  # note: no salt.minion_config line here... if you include it here, it will fail
  salt.install_master = false
  salt.no_minion = false
  salt.install_type = "git"
  salt.install_args = "v2015.5.1"
end  # salt provisioner

machine.vm.provision "shell",
  inline: "sudo cp /vagrant/salt/configs/minion_local /etc/salt/minion && sudo service salt-minion restart"

```

**Comment by babilen** [comment]
This is rather unfortunate as, for example, Debian jessie sets `PermitRootLogin without-password` in its `/etc/ssh/sshd_config` which prevents @nmadhok's suggestion from working.

Is there a way to utilise sudo in the copy process ?

**Comment by nmadhok** [comment]
@babilen I would recommend using a custom box image

**Comment by babilen** [comment]
@nmadhok Well, we are, but I am hesitant to change that globally in our vagrant cloud, simply because the provisioner is buggy. I'd rather recommend that people don't upgrade their vagrant installation from 1.7.2 yet.

**Comment by nmadhok** [comment]
@babilen I agree

**Comment by FlorianLoch** [comment]
Can confirm this one too, trusty64 seems to restrict root login too...

**Comment by davidkarlsen** [comment]
Any chance of getting this fixed for 1.7.5?

**Comment by nkantar** [comment]
Echoing @davidkarlsen's question. Any other info, ETA, etc.?

**Comment by jberends** [comment]
Worked around this permission problem as well by following @CharlesLovely comment by 'manually' copying over the minion config file and commenting out the config.minion_config

    config.vm.provision :salt do |salt|
        [...]
        #salt.minion_config = "saltstack/etc/minion"
        [...]
    end

    config.vm.provision "shell",
        inline: "sudo cp /etc/salt/minion{,-dist} && sudo cp /vagrant/saltstack/etc/minion /etc/salt/minion && sudo service salt-minion restart"

**Comment by onedeveloper** [comment]
I'm on the same boat, tried to use @jberends solution but it comes to a point where the highstate just hangs. My guess salt provisioner is trying to run with default minion config before copying the actual minion config file.

for reference 

Project structure:
$ tree
.
├── Vagrantfile
└── salt
    ├── minion
    └── roots
        ├── top.sls
        └── webserver.sls

2 directories, 4 files


Vagrantfile:
# -*- mode: ruby -*-
# vi: set ft=ruby :

VAGRANTFILE_API_VERSION = "2"

# bug with vagrant + salt provisioner
# see => https://github.com/mitchellh/vagrant/issues/5973
$bypass = <<SCRIPT
sudo cp /etc/salt/minion{,-dist} && sudo cp /vagrant/salt/minion /etc/salt/minion && sudo service salt-minion restart
SCRIPT

Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
  config.vm.define "web", primary: true do |web|
    web.vm.box = "ubuntu/trusty64"
    web.vm.network "forwarded_port", guest: 80, host: 8080
    web.vm.network "forwarded_port", guest: 443, host: 8443
    web.vm.synced_folder "salt/roots/", "/srv/salt/"
    web.vm.provision :salt do |salt|
      # salt.minion_config = "salt/minion" # bypassing normal minion configuration
      salt.run_highstate = true
      salt.bootstrap_options = "-P"
    end
    web.vm.provision "shell", inline: $bypass
  end
end

minion:
master: localhost
state_top: top.sls
file_client: local

tops.sls:
base:
  '*':
webserver

webserver.sls:
apache2:
  pkg:
    - installed

**Comment by pcn** [comment]
I've given up on 1.7.3 and 1.7.4.  Something regressed after 1.7.2.

**Comment by onedeveloper** [comment]
@pcn was it broken before 1.7.2 ?

**Comment by chrismcmacken** [comment]
I ran into the issue on 1.7.2, ended up using @amchoukir s workaround

**Comment by FlorianLoch** [comment]
I returned to 1.7.2 and can provision again... next timr I check the issues before blindly updating my dev box

**Comment by onedeveloper** [comment]
@chrismcmacken did you put that after or before the salt provisioner calls? could you paste your minion config and Vagrant File ?

**Comment by onedeveloper** [comment]
could someone maybe diff what changed in the provisioner between 1.7.2, 1.7.3 and 1.7.4 ?

**Comment by chrismcmacken** [comment]
@onedeveloper Basically you're just swapping from a salt provisioner to the shell provisioning and using that to install and run salt. You can go ahead and remove your salt provisioning block altogether. This also assumes that your salt minion config is in `salt/config/minion`, just update the `cp` line in the shell script if it's not.

add this line to Vagrantfile:
`config.vm.provision :shell, path: "salt_provision.sh", privileged: false`

salt_provision.sh
```#!/bin/bash

start_seconds="$(date +%s)"

if [ ! -d "/etc/salt" ]; then
    echo "Bootstrap salt"
    wget -O salt_bootstrap.sh https://bootstrap.saltstack.com
    sudo sh salt_bootstrap.sh -P
    sudo cp /vagrant/salt/config/minion /etc/salt/minion

    echo "Restart salt-minion"
    sudo salt-minion -d
    sudo service salt-minion restart
fi

echo "Executing salt highstate (provisionning)"
sudo salt-call state.highstate

end_seconds="$(date +%s)"
echo "-----------------------------"
echo "Provisioning complete in "$(expr $end_seconds - $start_seconds)" seconds"
```

minion:
`file_client: local`

**Comment by davidkarlsen** [comment]
It broke in 1.7.3. Some UNC path changes AFAIR. Referenced in this issue thread. Work around is to use vagrant to install minion (which gives functionality to select install method and version etc). But copy minion config and call highstate with shell. Not elegant but works.

**Comment by nkantar** [comment]
My solution is based on @amchoukir and @chrismcmacken's comments, and seems to be working fine with 1.7.4 and trusty64 (via VirtualBox).

```
config.vm.provision :salt do |salt|
  salt.bootstrap_options = "-P"
  salt.verbose = true
end

config.vm.provision "shell",
  inline: "sudo cp /etc/salt/minion{,-dist} && sudo cp /vagrant/salt/minion /etc/salt/minion && sudo salt-call --local state.highstate"
```

The important bit is the two omitted lines in the first block.

(Note: My `minion` file is located in `[repo]/salt/`.)

**Comment by onedeveloper** [comment]
@nkantar Thank you! and thanks everybody else that help figure out a workaround.

**Comment by virgofx** [comment]
Any insight into when this will be addressed?

**Comment by johntron** [comment]
I think there are two different ways of getting a `minion` file into the correct directory:
1. Letting [salt-bootstrap.sh do this](https://github.com/saltstack/salt-bootstrap/blob/stable/bootstrap-salt.sh#L4974) by calling `config_salt()`.
2. Have [vagrant's salt provisioner do this](https://github.com/mitchellh/vagrant/blob/master/plugins/provisioners/salt/provisioner.rb#L347).

#1 is not being run, because [vagrant installs using the `stable` branch of salt-bootstrap](https://github.com/mitchellh/vagrant/blob/master/plugins/provisioners/salt/bootstrap-salt.sh#L7), which [defaults to installing the stable version of salt](https://github.com/saltstack/salt-bootstrap/blob/develop/bootstrap-salt.sh#L228), and this branch of code does not call `config_salt()` – [only installing from git does](https://github.com/saltstack/salt-bootstrap/blob/develop/bootstrap-salt.sh#L1906).

#2 is failing, because the provisioner lacks permissions to the privileged `/etc/salt` directory.

**Comment by jrosskopf** [comment]
I'm not sure, why the provisioner has to even copy the `minion` config into the restricted `/etc/salt`. It would also be possible to copy it to an arbitrary place and run the salt call like

`sudo salt-call -c /vagrant/salt/ --local state.highstate`

**Comment by ddxgz** [comment]
Confirmed too.
Host: trusty64
box: trusty64
virtualbox: 4.3
vagrant: 1.7.4

**Comment by johnkeates** [comment]
Same problem here, downgrading to 1.7.2. again.

**Comment by johntron** [comment]
@ddxgz @johnkeates try giving my fix a +1: https://github.com/mitchellh/vagrant/issues/6073

**Comment by obestwalter** [comment]
> I'm not sure, why the provisioner has to even copy the minion config into the restricted /etc/salt. It would also be possible to copy it to an arbitrary place and run the salt call like

> ``sudo salt-call -c /vagrant/salt/ --local state.highstate``

:+1:  for arbitrary place but  :-1:  for ''/vagrant/salt/'' - I would not use the mounted vagrant folder at all as this contains the users files and should be left untouched. Does vagrant have a generic tmp folder that has wide open permissions and works on all supported Osses already?

**Comment by onedeveloper** [comment]
I have 2 comments regarding this.

1. you can mount an extra directory for you to call i.e. /config/salt
mapped to a directory outside of /vagrant

2. this is my only concern. How do you specify that for calls to vagrant
provision?

**Comment by obestwalter** [comment]
sorry - I had originally removed those concerns, thinking they are not valid, if we just have a convention for this (readded them and will not do so much aggressive editing in the future).

> 1. you can mount an extra directory for you to call i.e. /config/salt mapped to a directory outside of /vagrant

I think this would not be a good way to deal with this. It should be transparent to the user. You should be able to say where your configs/keys/etc. are, just like now and then they it will get applied during provisioning. So my idea about the convention is actually stupid, because completely unnecessary.

salt provisioner config looks like this atm:
    salt.minion_config = "path/to/my/minionconfig"
    salt.minion_key = "path/to/my//minion.pem" 
    salt.minion_pub = "path/to/my//minion.pub"

So, as a user I provide the paths to the files and how they will get applied is not my problem. 

This (untested) fix by @johntron will hopefully address this.

**Comment by dragon788** [comment]
Personally we add our custom salt minion configs to the /etc/salt/minion.d/ directory. This one is checked by default by the salt-minion as additional configuration, so anything it finds overrides the default /etc/salt/minion file (which is almost always what you want anyways). I would never suggest editing/replacing the default file, because package upgrades can clobber that file, while the minion.d directory with your custom file will stay intact.

**Comment by obestwalter** [comment]
> Personally we add our custom salt minion configs to the /etc/salt/minion.d/ directory.

That is good advice for real live situations, but this is vagrant and if I am not missing something, all you can do is order the provisioner to overwrite the minion config with a file you provide in this setting: ``salt.minion_config = "path/to/my/minionconfig"`` - amiright?

**Comment by johntron** [comment]
@obestwalter correct, and vagrant just copies the minion file to /tmp before running the bootstrap - it's the bootstrap that puts it in the proper location. This issue was opened, because vagrant doesn't have permissions to put things directly into /etc/salt (minion or minion.d/ as @dragon788 suggested). A shell provisioner could do this, but I think it's best to just use salt-bootstrap's default behavior for this sort of thing.

@dragon788 you're right about updating packages potentially clobbering the minion file, but I've never had this problem in my development environments, and in production I've never used masterless provisioning.

**Comment by darkn3rd** [comment]
I just started playing with salt, and had the same issue.  I noticed that there was a prior issue #4032 maybe related to this.  I guess the workaround (guessing at the behavior) would be to scp it to a /tmp directory, then sudo cp it to appropriate place.  Has anyone did a pull request related to this issue?

**Comment by johntron** [comment]
@darkn3rd yeah, you could use a shell provisioner to `sudo cp` it to the appropriate directory  – from a mounted/shared folder or via a file provisioner – but this will require that vagrant has `nopasswd` sudo privileges, which may not be the case for everyone.

I think the best solution is to get [my PR](https://github.com/mitchellh/vagrant/pull/6073) merged, then everyone can just rely on salt-bootstrap [like I do](https://github.com/mitchellh/vagrant/pull/6073#issuecomment-129526388).

**Comment by kaidokert** [comment]
Ugh. I just needed a quick salt sandbox and here is what i ended up doing, as provisioner salt.* options turn out to be pretty broken and useless currently.

```
Vagrant.configure(2) do |config|
  config.vm.box = "ubuntu/trusty64"
  config.vm.network "private_network", ip: "10.0.0.2"
  config.vm.synced_folder "salt/roots", "/srv/salt/"
  config.vm.synced_folder "salt/host", "/host"
  config.vm.provision "shell", inline: "sudo mkdir -p /etc/salt /etc/salt/{master.d,minion.d}"
  config.vm.provision "shell", inline: "sudo cp -a /host/minion.d /etc/salt/"
  config.vm.provision "shell", inline: "sudo cp -a /host/master.d /etc/salt/"
  config.vm.provision :salt do |salt|
    salt.verbose = true
    salt.install_command = '-P -M -K' # http://goo.gl/m5TUnO
  end 
  config.vm.provision "shell", inline: "sudo salt-key -y -A"
end

```

**Comment by zundra** [comment]
Best/easiest work around I've found so far is to downgrade to 1.7.2.

**Comment by blueyed** [comment]
What seemed to work for me:

Don't set `salt.minion_config`, and then use `minion.d` instead:

    config.vm.synced_folder "minion.d", "/etc/salt/minion.d"

This requires `salt.always_install = true`, too - which seems to get skipped without `salt.minion_config`.

**Comment by tehmaspc** [comment]
this is the 2nd issue w/ salt provisioner today; i think @zundra's advice is best for now till vagrant 1.8. or some patch fix to 1.7.x.

substituting the salt provisioner w/ a shell provisioner doing the same thing is not a valid maintainable fix - I could do this and have for other provisioners in the past - but ain't nobody got time for that.

**Comment by amchoukir** [comment]
@tehmaspc out of curiosity why do you think a shell provisionner is not a maintainable solution. Vagrant caters to type 2 hypervisors, I.e the hypervisors runs as an application on a host OS  and allow to run guest operating systems. For type 1 hypervisors, the hypervisor runs bare metal not on top of a host operating system. Examples would be VMWare ESXi, Citrix Xenserver and so on. Type 1 hypervisor are the heavy duty case running hundred if not thousands of VMs and Salt would most likely be installed via a shell script or through a gold image/template.

**Comment by edvinasme** [comment]
Excerpt from a working Vagrant file (with a workaround applied):

```
  config.vm.provision :salt do |salt|
    #salt.minion_config = "config/minion" # vagrant issue #5973
    salt.run_highstate = false # vagrant issue #5973
    salt.log_level     = "warning" # all,garbage,trace,debug,info,warning

    # use older version of salt in order to prevent rabbitmq bug:
    # https://github.com/saltstack/salt/issues/25118
    salt.install_type = "git"
    salt.install_args = "v2015.5.0"
  end

  # vagrant issue #5973
  config.vm.provision "shell",
    inline: "sudo cp /etc/salt/minion{,-dist} && sudo cp /vagrant/config/minion /etc/salt/minion && sudo salt-call --local state.highstate"

```

`minion` configuration file is kept in `config/minion` (directory relative to Vagrantfile).

Is this issue still affecting Salt 2015.5.5 or 2015.8.0RCs?

**Comment by ev0rtex** [comment]
@edvinasme This bug exists regardless of the salt version.

If it helps anyone else, here's what I did. In my case I'm passing pillar data along in my Vagrantfile so I can't just make it provision using a shell command. Instead I make it copy the minion config to `/tmp/minion` and then pass in the `-c /tmp` bootstrap option:

```ruby
    conf.vm.provision :file, source: "provision/minion", destination: "/tmp/minion"
    conf.vm.provision :salt do |salt|
        # Custom pillar
        salt.pillar("user_options" => vmopts['options'].presence || {})
        salt.pillar("user_packages" => vmopts['packages'].presence || {})
        salt.pillar("timezone" => zone)

        # Salt options
        salt.minion_config = "provision/minion"
        salt.run_highstate = true
        salt.install_type = "git"
        salt.install_args = "v2015.8.0rc3"
        salt.bootstrap_options = "-P -c /tmp"
        salt.colorize = true
        salt.verbose = true
        salt.log_level = "info"
        # salt.always_install = true
    end
```

**Comment by joshughes** [comment]
@tehmaspc I can confirm vagrant 1.7.4 and Virtualbox work with the above fix... which looks like the following for me..

```ruby 
  config.vm.provision :file, source: "./.vagrant-salt/minion", destination: "/tmp/minion"
  # Provision VM with current formula, in masterless mode
  config.vm.provision :salt do |salt|
    salt.minion_config = "./.vagrant-salt/minion"
    salt.run_highstate = true
    salt.install_type = 'git v2015.5.5'
    salt.bootstrap_options = "-P -c /tmp"
    salt.colorize = true
    salt.verbose = true
  end
```

**Comment by bemeyert** [comment]
Hi all, I stumbled over this issue because my setup for minion boxes (pre Vagrant 1.7.2) was no longer working with 1.7.4 too. After some gruesome debugging (eyes still hurting) I found out that vagrant uploads the keys and configuration only to the "tmpdir" (in Linux this is /tmp) as user "vagrant". So I gave the bootstrap script some options via `salt.bootstrap_options`: `-F -c /tmp/`. This worked at the end. So my setup looks like this:
```ruby
[...]
$minion_script = <<EOF
yum remove -y salt\*
echo $(hostname -f) > /etc/salt/minion_id
EOF
[...]
        minionc6.vm.provision "shell",
            inline: $minion_script
        minionc6.vm.provision :salt do |salt|
            salt.bootstrap_options = '-F -c /tmp/'
            salt.minion_config = "provision/minion_config"
            salt.minion_key = "provision/keys/minion.pem"
            salt.minion_pub = "provision/keys/minion.pub"
            salt.colorize = true
            salt.log_level = "info"
            salt.verbose = true
        end
[...]
```
So I think that either the documentation should state the usage of the switches to bootstrap when provisioning keys and config or vagrant could do it by itself.

**Comment by robertwe** [comment]
@bemeyert works for me. Many thx for sharing it.

**Comment by bemeyert** [comment]
@robertw Glad I could help ;)

**Comment by bsuh** [comment]
# TL;DR: Add `salt.bootstrap_options = '-F -c /tmp/ -P'` to your Vagrantfile

# Salt bootstrap options explanations
```
  Options:
  -c  Temporary configuration directory
  -F  Allow copied files to overwrite existing(config, init.d, etc)
  -P  Allow pip based installations. On some distributions the required salt
      packages or its dependencies are not available as a package for that
      distribution. Using this flag allows the script to use pip as a last
      resort method. NOTE: This only works for functions which actually
      implement pip based installations.
```
https://github.com/saltstack/salt-bootstrap/blob/b1624cc14a91f21abab1da0fe1ef23d5a95d2ef1/bootstrap-salt.sh#L245

# WTF is going on
There's multiple issues messing up provisioning.

## #6029 : Bootstrapping Salt Fails Due to Salt-Boostrap Update
A salt-bootstrap script update makes it so that `-P` needs to be passed to the bootstrap script for PIP based installations to work.

## #6011 : Regression: Salt provisioning fails when box is already provisioned
Vagrant no longer passing in `-F -c <temporary config dir>` to bootstrap options. Committed https://github.com/mitchellh/vagrant/commit/da67824a53ba0e7c2db1a9b8e19141c1ffd07bfc#diff-c4526b51e4e1b1f3f6bff789181cce2bR107 11 days before v1.7.4
and fixed on ff941ef4e8b8b8b9f90c9cf270cb6dafd0593407 post v1.7.4.

You can see that the added conditional is written incorrectly as `!@machine.config.vm.communicator == :winrm` instead of `@machine.config.vm.communicator != :winrm` or `!(@machine.config.vm.communicator == :winrm)`

If salt is already installed, then configuration mode fails without temporary configuration directory specified through `-c`
```
if [ "$_CONFIG_ONLY" -eq $BS_TRUE ] && [ "$_TEMP_CONFIG_DIR" = "null" ]; then
    echoerror "In order to run the script in configuration only mode you also need to provide the configuration directory."
    exit 1
fi
```
https://github.com/saltstack/salt-bootstrap/blob/b1624cc14a91f21abab1da0fe1ef23d5a95d2ef1/bootstrap-salt.sh#L381

If salt isn't already installed, then `bootstrap-salt.sh` uses the temporary configuration directory to copy over configuration files. But since the temporary configuration directory isn't passed in, no configuration files are copied over.

## #5973 : Salt provisioner fails to upload minion config

During highstate, vagrant tries to copy the minion config file to `/etc/salt/`. Committed https://github.com/mitchellh/vagrant/commit/74d22069d41b35bae50e32c1c68a92a4958afebb#diff-c4526b51e4e1b1f3f6bff789181cce2bR308 March 11, 2014. The reason was for base boxes with salt already installed, provisioning skipped installing salt and did not copy over minion configuration files (no longer true).

Wait March 11, 2014? Why did it take so long for permission errors to occur?!
```
vagrant@saltminion1:~$ ls -l /etc/salt/
total 60
-rw-r--r-- 1 vagrant vagrant 23480 Sep  2 19:23 minion
drwxr-xr-x 2 root    root     4096 Sep  2 19:13 minion.d
-rw-r--r-- 1 vagrant vagrant 25854 Jul  7 20:42 minion.dpkg-dist
drwx------ 3 root    root     4096 Sep  2 19:13 pki
```
So if `bootstrap-salt.sh` receives a temporary configuration directory, it copies over the files with the same permissions, and it is writeable by the `vagrant` user. So because of #6011, #5973 is exposed as a bug.

#6073 : Stop trying to upload minion config to privileged directory
This (unmerged as of now) pull request stops unneeded extra copy to `/etc/salt/`, since provisioning with salt already installed, does copy over minion config files.

# Summary
#6011 causes #5973 to be revealed. #6073 pull request is awaiting merge to fix #5973.
#6011 is fixed in ff941ef4e8b8b8b9f90c9cf270cb6dafd0593407
#6029 is not fixed and no pull requests for it.

# Bonus
Figuring out how to use git master took me all day to figure out, so I'll document it here. It could be a better option when a bug is fixed on master than waiting for the next release. I found something on Google about using `VAGRANT_INSTALLER_EMBEDDED_DIR`, but it ended up crashing with plugins and I couldn't figure out how to deal with it. Eventually I just tried installing the master as a gem into the Ruby environment embedded inside the installed vagrant.

Replace `/opt/vagrant/` with your vagrant installation directory as necessary.
```
chown -R <user> /opt/vagrant/
cd
git clone https://github.com/mitchellh/vagrant.git
cd vagrant
GEM_HOME=/opt/vagrant/embedded/gems/ /opt/vagrant/embedded/bin/gem install bundler -v 1.10.5
GEM_HOME=/opt/vagrant/embedded/gems/ /opt/vagrant/embedded/gems/bin/bundle install
GEM_HOME=/opt/vagrant/embedded/gems/ /opt/vagrant/embedded/gems/bin/rake install
```

**Comment by johntron** [comment]
> causes #5973 to be revealed. #6073 pull request is awaiting merge to fix #5973.
#6011 is fixed in ff941ef
#6029 is not fixed and no pull requests for it.

Exactly what I found as well. 

I wasn't sure how #6029 should be fixed. If we include `-P` for all systems, some may not work due to special cases. The system packages (e.g. Distro packages from APT) may have required special build steps or patches, so installing from pip may not work. Conversely, a system package may not be available in the right version.

Maybe we should include `-P` by default, since this will be most likely to succeed, and let the user explicitly disable it with an option?

**Comment by tehmaspc** [comment]
@bsuh - LMAO - great post formatting.

**Comment by mitchellh** [comment]
Fixed by #6073

**Comment by dwickwire** [comment]
:+1:   Thanks!

**Comment by kevich** [comment]
Hooray! Thaks! Hope we will get new 1.8.0 release soon

**Comment by tehmaspc** [comment]
Thank You!

**Comment by obestwalter** [comment]
@mitchellh Wow - lot's of salt provisioner merges today- thanks a lot!

**Comment by AndrewPashkin** [comment]
@bsuh :thumbsup:

**Comment by BrianGallew** [comment]
For those of you who aren't using bleeding-edge release, there's a much more trivial workaround:

```
      node.vm.provision 'shell',
          inline: 'mkdir /etc/salt;chgrp vagrant /etc/salt;chmod g+w /etc/salt'
```

**Comment by dwickwire** [comment]
Can we get an update on when a new release with this fix may land?  We're just trying to plan accordingly and having an idea would help.

Thanks

**Comment by Jetlum** [comment]
Got the same error then I just wrote the commands on cmd instead of git console and it worked.
